src/main/java/uk/gov/hmcts/ccd/domain/service/casedataaccesscontrol/RoleToAccessProfilesMappingsGenerator.java

```diff
@@ -0,0 +1,126 @@
+import uk.gov.hmcts.ccd.domain.model.casedataaccesscontrol.AccessProfile;
+import uk.gov.hmcts.ccd.domain.model.definition.AccessControlList;
+import uk.gov.hmcts.ccd.domain.model.definition.CaseTypeDefinition;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static uk.gov.hmcts.ccd.data.caseaccess.GlobalCaseRole.CREATOR;
+
+@Component
+public class RoleToAccessProfilesMappingsGenerator {
+
+    protected static final String IDAM_PREFIX = "idam:";
+    private static final String CASE_ROLE_ID_REGEX = "^(\\[[A-Za-z]+\\])$";
+
+    private final CaseRoleRepository caseRoleRepository;
+
+    public RoleToAccessProfilesMappingsGenerator(@Qualifier(CachedCaseRoleRepository.QUALIFIER)
+                                                     CaseRoleRepository caseRoleRepository) {
+        this.caseRoleRepository = caseRoleRepository;
+    }
+
+    public List<AccessProfile> generate(CaseTypeDefinition caseTypeDefinition) {
+
+        Set<String> caseRoles = generateCaseRoles(caseTypeDefinition);
+        Set<String> idamRoles = generateIdamRoles(caseTypeDefinition);
+
+        List<AccessProfile> accessProfiles = new ArrayList<>();
+        accessProfiles.addAll(createAccessProfiles(caseTypeDefinition.getId(), caseRoles, false));
+        accessProfiles.addAll(createAccessProfiles(caseTypeDefinition.getId(), idamRoles, true));
+
+        return accessProfiles;
+    }
+
+    private Set<String> generateIdamRoles(CaseTypeDefinition caseTypeDefinition) {
+        Set<String> idamRoles = new HashSet<>();
+
+        idamRoles.addAll(getIdamRolesFromAcls(caseTypeDefinition.getAccessControlLists()));
+        idamRoles.addAll(getIdamRolesFromAcls(eventAcls(caseTypeDefinition)));
+        idamRoles.addAll(getIdamRolesFromAcls(stateAcls(caseTypeDefinition)));
+        idamRoles.addAll(getIdamRolesFromAcls(caseFieldAcls(caseTypeDefinition)));
+        idamRoles.addAll(getIdamRolesFromAcls(caseFieldComplexAcls(caseTypeDefinition)));
+
+        return idamRoles;
+    }
+
+    private Set<String> generateCaseRoles(CaseTypeDefinition caseTypeDefinition) {
+        final Set<String> validCaseRoles = caseRoleRepository.getCaseRoles(caseTypeDefinition.getId());
+        boolean containsCreator = validCaseRoles.contains(CREATOR.getRole());
+
+        Set<String> caseRoles = new HashSet<>();
+        if (!containsCreator) {
+            caseRoles.add(CREATOR.getRole());
+        }
+        caseRoles.addAll(getCaseRolesFromAcls(caseTypeDefinition.getAccessControlLists()));
+        caseRoles.addAll(getCaseRolesFromAcls(eventAcls(caseTypeDefinition)));
+        caseRoles.addAll(getCaseRolesFromAcls(stateAcls(caseTypeDefinition)));
+        caseRoles.addAll(getCaseRolesFromAcls(caseFieldAcls(caseTypeDefinition)));
+        caseRoles.addAll(getCaseRolesFromAcls(caseFieldComplexAcls(caseTypeDefinition)));
```

Comment:
I'm confused. If you are calling Def Store to get the list of case roles for a case type, why then you are doing the work to extract them from the ACLs?